src/WebJobs.Script/Host/BlobLeaseManager.cs

```diff
@@ -0,0 +1,173 @@
+
+            // Renew the lease three seconds before it expires
+            _renewalInterval = leaseTimeout.Add(TimeSpan.FromSeconds(-3));
+
+            // Attempt to acquire a lease every 5 seconds
+            _leaseRetryInterval = TimeSpan.FromSeconds(5);
+
+            _timer = new Timer(o => AcquireOrRenewLeaseAsync().ContinueWith(t => { _traceWriter.Error($"Error processing host blob lease: {t.Exception}"); }, TaskContinuationOptions.OnlyOnFaulted),
+            null, TimeSpan.Zero, _leaseRetryInterval);
+        }
+
+        public event EventHandler HasLeaseChanged;
+
+        public bool HasLease => _leaseId != null;
+
+        public string LeaseId
+        {
+            get
+            {
+                return _leaseId;
+            }
+            private set
+            {
+                string previousId = _leaseId;
+                _leaseId = value;
+
+                if (string.Compare(previousId, _leaseId, StringComparison.OrdinalIgnoreCase) != 0)
+                {
+                    OnHasLeaseChanged();
+                }
+            }
+        }
+
+        private void OnHasLeaseChanged() => HasLeaseChanged?.Invoke(this, EventArgs.Empty);
+
+        public static async Task<BlobLeaseManager> CreateAsync(string accountConnectionString, TimeSpan leaseTimeout, string hostId, TraceWriter traceWriter)
+        {
+            if (leaseTimeout.TotalSeconds < 15 || leaseTimeout.TotalSeconds > 60)
+            {
+                throw new ArgumentOutOfRangeException(nameof(leaseTimeout), $"The {nameof(leaseTimeout)} should be between 15 and 60 seconds");
+            }
+
+            ICloudBlob blob = await GetLockBlobAsync(accountConnectionString);
+            var manager = new BlobLeaseManager(blob, leaseTimeout, hostId, traceWriter);
+            return manager;
+        }
+
+        public static BlobLeaseManager Create(string accountConnectionString, TimeSpan leaseTimeout, string hostId, TraceWriter traceWriter)
+        {
+            return CreateAsync(accountConnectionString, leaseTimeout, hostId, traceWriter).GetAwaiter().GetResult();
+        }
+
+        private async Task AcquireOrRenewLeaseAsync()
+        {
+            try
+            {
+                if (HasLease)
+                {
+                    await _lockBlob.RenewLeaseAsync(new AccessCondition { LeaseId = LeaseId });
+                    _traceWriter.Verbose("Host lock lease renewed.");
+                }
+                else
+                {
+                    LeaseId = await _lockBlob.AcquireLeaseAsync(_leaseTimeout, _hostId);
+                    _traceWriter.Info($"Host lock lease acquired by host ID '{_hostId}'.");
+
+                    // We've successfully acquired the lease, change the timer to use our renewal interval
+                    _timer.Change(_renewalInterval, _renewalInterval);
+                }
+            }
+            catch (StorageException exc) when (exc.RequestInformation.HttpStatusCode == 409)
```

Comment:
>catch (StorageException exc) when (exc.RequestInformation.HttpStatusCode == 409) [](start = 12, length = 80)

Should we remove the LeaseId if any exception is thrown here? Or is that handled some other way?

Example -- the primary instance can't connect for some reason and times out trying to renew. It will still have a LeaseId even though it'll time out in a few seconds. There may be the possibility of two hosts with HasLease = true for a short time.